### ISSUE TITLE
[Feature Request] Fix deposition rate specified in time units 

### ISSUE THREAD

**Issue by tomasfbouvier** [lammps/lammps#4158]
**Summary**

The fix deposition command currently only allows to specify the rate at which atoms are inserted in timesteps. Is it possible to add an alternative parameter to specify the rate as a time delay instead?

**Detailed Description**

The fix deposition command insterts atoms every M timesteps. When using a fixed timestep one can control the time delay between insertions. However when using adaptative timestep the number of timesteps required for a time interval varies and therefore one can only approximately control the rate of deposition. Many depositions involve high energy particles for which adt is needed. In order to gain full control one needs to combine the previous commands with halting and for loops. It would be much simpler to have a parameter to control the time delay between depositions directly within the fix command.

**Comment by abkein** [comment]
I want to try to make it to work, but I need some guidance:
- Do we need to keep backward compatibility here? I.e. can we make 5-th to be specified in time units and not number of steps? Or just add a flag signaling that the argument is specified in time units?

Time management:
1. When reneighboring happens, we can calculate the step of `next_reneighbor` with the current timestep value; however,  timestep value can be changed between two consequential reneighborings.
2. We can add the `end_of_step` function and at the end of each step subtract timestep value from the remaining time, and in case `remaining_time <= 0` set `next_reneighbor = update->ntimestep +1`, then after deposition restore `remaining_time = freq`.

What approach will be better?
In case of second approach why don't allow deposition rate to be a variable?

**Comment by akohlmey** [comment]
> I want to try to make it to work, but I need some guidance:
> 
> * Do we need to keep backward compatibility here? I.e. can we make 5-th to be specified in time units and not number of steps? Or just add a flag signaling that the argument is specified in time units?

Backward compatibility is and has always been a high priority. There is precedence for having a flag to change how the unit of parameters is interpreted with the "units box/lattice" parameters in many places for length units. The complication here is that "nfreq" is currently an integer parameter and specifying a time requires relaxing the check since that would require a floating point number. One possible way to deal with this would be to parse this argument  **after** all the other argument parsing. So you could add a "depunit" argument which could be "step" or "time" and then you do the setting of "nfreq" dependent on the depunit argument (which should become an integer flag with two (local) enumerator values) and for "time", you would set a new member, tfreq, from the argument and from that and `update->dt` you then compute nfrequ.

> Time management:
> 
> 1. When reneighboring happens, we can calculate the step of `next_reneighbor` with the current timestep value; however,  timestep value can be changed between two consequential reneighborings.
> 2. We can add the `end_of_step` function and at the end of each step subtract timestep value from the remaining time, and in case `remaining_time <= 0` set `next_reneighbor = update->ntimestep +1`, then after deposition restore `remaining_time = freq`.
> 
> What approach will be better? In case of second approach why don't allow deposition rate to be a variable?

Neither. For the most part you can probably keep assuming that nfreq remains constant. When fix dt/reset changes the timestep, the `Fix::reset_dt()` method is called, so that method should be added to fix deposit and then nfreq recomputed from tfreq and all timestep related parameters reset as needed, e.g. next_reneighbor.

**Comment by abkein** [comment]
But these lines
```C++
// setup_pre_exchange
if (ninserted < ninsert) next_reneighbor = nfirst + ((update->ntimestep - nfirst)/nfreq)*nfreq + nfreq;
else next_reneighbor = 0;

// pre_exchange
if (rateflag) offset = (update->ntimestep - nfirst) * update->dt * rate;
```
Assuming timestep value remains constant would lead to an increasing error in `next_reneighbor` (in case setup is called several times) and `offset` calculation, which would increase with `update->ntimestep - nfirst`.

We either need to keep track of timestep value changes or (and better I think) introduce a counter `ncalled` of how many times the deposition has been performed, and change `nfirst` to `nprev`. And rewrite it like this
```C++
// setup_pre_exchange
if (depunitflag != 0) nfreq = static_cast<int>(tfreq / update->dt);
if (ninserted < ninsert) next_reneighbor = nprev + nfreq;
else next_reneighbor = 0;

// pre_exchange
if (rateflag)
  if (depunitflag == 0) offset = update->dt * ncalled * nfreq * rate;
  else offset = ncalled * tfreq * rate;
```
And in restart save both `ncalled` and `nprev`.

Just in case
```C++
void FixDeposit::reset_dt() 
{
  if (depunitflag) {
    nfreq = static_cast<int>(tfreq / update->dt);
    next_reneighbor = nprev + nfreq;
  }
}
```

**Comment by akohlmey** [comment]
@abkein I don't understand the logic of what you are arguing here.

When you are in insert by time mode, every time `Fix::reset_dt()` is called, you have to treat it as if you are starting the depositing fresh with a new nfreq and also thus also the next_reneighbor step has to be the next timestep, just like at the beginning, i.e. combine what is done in the constructor and in `Fix::setup_pre_exchange()`. For as long as the number of inserted particles is less than the requested number, that is. 

Basically, you have to treat the simulation as if it was broken down into pieces and each would be with a different insertion rate.

**Comment by akohlmey** [comment]
What I would suggest you do is to write a minimal fix for debugging, where you change dt manually from a list of timesteps and corresponding dt values, which would then function as a mockup version of fix dt/reset. Then you can set up a dummy system (something similar like the melt example, but with just a couple of atoms) where you insert more atoms. That way, you can emulate what happens in a real application of fix dt/reset, but with much less effort and much faster. With that test harness, you can then first monitor the values of the various parameters and then check if you get the expected behavior from your implementation. Since you know exactly from your mockup fix how long segments with which dt and insertion frequency you have, you can easily confirm that everything is running as expected.
The behavior should be the same as if you had the same simulation where you have several run commands and change dt after each and replace the fix deposit command with the corresponding settings for the new dt.
